middleware/sync.js

```diff
@@ -36,51 +48,58 @@ const navRenderSub = new Rx.SerialDisposable();
  * The middleware itself - passes the queries to the application server, which
  * will make necessary calls to the API
  */
-const getSyncMiddleware = routes => store => next => action => {
-	if (action.type === LOCATION_CHANGE ||  // client nav
-		action.type === '@@server/RENDER' ||
-		action.type === 'LOCATION_SYNC') {
-		const dispatchApiRequest = bindActionCreators(apiRequest, store.dispatch);
+const getSyncMiddleware = routes => store => {
+	/**
+	 * We want to make sure there is always and only one data-fetching subscription
+	 * open at a time, so we create an empty subscription that is scoped to the
+	 * request. When a new fetch happens, this subscription will get replaced with
+	 * a new subscription.
+	 */
+	const actions = bindActionCreators(
+		{ apiRequest, apiSuccess, apiError, apiComplete, locationSync },
+		store.dispatch
+	);
 
-		const location = action.payload;
-		const activeQueries$ = activeRouteQueries$(routes, { location })
-			.delay(0)  // needed in order for LOCATION_CHANGE to finish processing
-			.filter(queries => queries);  // only emit value if queries exist
+	const doFetch = serialSubscription(
+		actions.apiSuccess,
+		actions.apiError,
+		actions.apiComplete
+	);
 
-		activeQueries$.subscribe(dispatchApiRequest);
-	}
+	return next => action => {
+		if (action.type === LOCATION_CHANGE ||  // client nav
```

Comment:
Something that would be helpful is to have more inline comments that describe what each block of code is doing. i.e., what is this return statement? or why is a locationSync happening on line 84? etc.